gh-113743: Make the MRO cache thread-safe in free-threaded builds

DinoV opened the pull request:
This makes _PyType_Lookup thread safe.

There's a couple of aspects to this:
1) Thread safety of the underlying cache.

For this a lock that is similar to Linux's seqlock is used.  This allows the reader to read the sequence using acquire semantics which ensures it sees the values that were written by the last time the lock was released.  It reads the sequence a 2nd time again with acquire semantics after having read the values and confirms the sequence has not changed, in which case the values read were valid.  

On an update if we have multiple readers who missed but looked up the same value only the first one in will perform the update.  It also appears the linux implementation (at least initially?) used a separate spin lock, we just use CAS on the sequence.

There's currently a sequence/lock per-cache entry, on 64-bit builds this is free due to the alignment of the existing cache entries.

2) Make mutation of `mro` and type members thread safe
Because the cache is racing against mutations to the types those mutations also need to be thread safe.  This involves adding critical sections around mutation of the type's dictionary (type_setattro), mutation of the mro, and mutation of the bases (which of course ultimately mutates the mro).  

This bubbles up to a surprising number of functions as `lookup_tp_mro` and both are returning borrowed references which need to be protected in case the type is modified.  In general I've used assertions to make sure the locks are already held by the time we reach the functions using them to push the locking into the chunkier functions.

There is a single mutex which is used to protect all type objects.  This is because sometimes we need to actually update the type hierarchy, and if were to recurse through sub-types and lock them individually we could end up releasing the outer type's lock and open it up to mutation in a way which is not currently possible. 

`Py_BEGIN_CRITICAL_SECTION` and `Py_END_CRITICAL_SECTION` are a little inflexible in that you can't return inside of a critical section, so for more complicated functions I've added `_unlocked` variants and wrapped those in the original name which acquires the lock.

These unlocked versions are also used elsewhere when we know that we have the lock held.  If we've already acquired the lock currently re-acquiring the lock immediately, even from the same thread, comes with some additional cost as we first spin and then need to go to the parking lot.  

Also `_PyType_GetMRO` and `_PyType_GetBases` are currently returning borrowed references which aren't safe.  Luckily these functions aren't currently exported, so I've switched them to returning owned references.  `_PyType_GetBases` actually had no consumers before this diff but now is used internally to typeobject.c.  `_PyType_GetMRO` was only used in the abc module, and even then it's usage is actually just re-implementation of `PyType_IsSubtype` so I've replaced it with a call to that function.

In the future we may be able to improve this.  Currently much of this locking is done just to make sure we don't end up with a dangling reference to mro/bases.  In some cases it may be okay to race against mutations and just have the program behave as if whatever check is applied to those happened before the mutation (e.g. consider the case of `PyType_IsSubtype`, whether the mutation happens right after we fetch MRO or right after we release doesn't really matter, as the type can change as soon as we've released the lock).  Therefore we may be able to eliminate some of this locking by having a mutation of mro/bases scheduling the final dec ref to happen via qsbr when when hit the next quiescent state.

<!-- gh-issue-number: gh-113743 -->
* Issue: gh-113743
<!-- /gh-issue-number -->

DinoV committed:
Make type cache thread safe with sequence lock

DinoV committed:
Make bases/mro accesses thread safe

colesbury commented on Include/internal/pycore_typeobject.h:
```diff
@@ -74,7 +79,7 @@ struct types_state {
 extern void _PyTypes_FiniTypes(PyInterpreterState *);
 extern void _PyTypes_Fini(PyInterpreterState *);
-
+void _PyTypes_AfterFork(void);
```
```suggestion
extern void _PyTypes_AfterFork(void);
```

I don't think it matters. Just for consistency.

colesbury commented on Include/internal/pycore_lock.h:
```diff
@@ -251,6 +251,7 @@ PyAPI_FUNC(void) _PyRWMutex_RUnlock(_PyRWMutex *rwmutex);
 PyAPI_FUNC(void) _PyRWMutex_Lock(_PyRWMutex *rwmutex);
 PyAPI_FUNC(void) _PyRWMutex_Unlock(_PyRWMutex *rwmutex);
 
+void _Py_yield(void);
```
```suggestion
extern void _Py_yield(void);
```

colesbury commented on Include/internal/pycore_critical_section.h:
```diff
@@ -236,6 +253,19 @@ _PyCriticalSection2_End(_PyCriticalSection2 *c)
+#ifdef Py_GIL_DISABLED
+
+static inline void
+_PyCriticalSection_AssertHeld(PyMutex *mutex) {
```
```suggestion
_PyCriticalSection_AssertHeld(PyMutex *mutex)
{
```

colesbury commented on Objects/typeobject.c:
```diff
@@ -4747,6 +4877,82 @@ is_dunder_name(PyObject *name)
+void _PyTypes_AfterFork() {
+#ifdef Py_GIL_DISABLED
+    struct type_cache *cache = get_type_cache();
+    for (int i = 0; i < 1 << MCACHE_SIZE_EXP; i++) {
```
For consistency with other loops over the cache:

```suggestion
    for (Py_ssize_t i = 0; i < (1 << MCACHE_SIZE_EXP); i++) {
```

colesbury commented on Objects/typeobject.c:
```diff
@@ -4747,6 +4877,82 @@ is_dunder_name(PyObject *name)
+
+#endif
+
+void _PyTypes_AfterFork() {
```
```suggestion
void
_PyTypes_AfterFork()
{
```

colesbury commented on Objects/typeobject.c:
```diff
@@ -4747,6 +4877,82 @@ is_dunder_name(PyObject *name)
+        } else {
+            _Py_yield();
+        }
+    }
```
Can you refactor out the seqlock code into a type and functions, possibly in `pycore_lock.h`?

colesbury commented on Objects/typeobject.c:
```diff
@@ -4747,6 +4877,82 @@ is_dunder_name(PyObject *name)
+        if (TYPE_CACHE_IS_UPDATING(prev)) {
+            // Someone else is currently updating the cache
+            _Py_yield();
+            prev = _Py_atomic_load_int32_relaxed(&entry->sequence);
```
Should be `int` (not `int32`) operations. (Alternatively, declare `sequence` as `int`).

colesbury commented on Objects/typeobject.c:
```diff
@@ -4747,6 +4877,82 @@ is_dunder_name(PyObject *name)
+        struct type_cache_entry *entry = &cache->hashtable[i];
+        int sequence = _Py_atomic_load_int_acquire(&entry->sequence);
+        if (TYPE_CACHE_IS_UPDATING(sequence)) {
+            // Entry was in the process of updating while forking, clear it...
```
I don't think this case should ever happen. In free-threaded builds, we will stop-the-world before forking so that a fork() doesn't happen in the middle of some code like this.

We're not doing that yet, but the GIL also ensures that we don't fork in the middle updating the type cache.

colesbury commented on Objects/typeobject.c:
```diff
@@ -4747,6 +4877,82 @@ is_dunder_name(PyObject *name)
+    if (entry->name == name &&
+        entry->value == value &&
+        entry->version == version_tag) {
+        // We reaced with another update, bail and restore previous sequence.
```
```suggestion
        // We raced with another update, bail and restore previous sequence.
```

colesbury commented on Objects/typeobject.c:
```diff
@@ -4747,6 +4877,82 @@ is_dunder_name(PyObject *name)
+
+#if Py_GIL_DISABLED
+
+#define TYPE_CACHE_IS_UPDATING(sequence) (sequence & 0x01)
```
```suggestion
#define TYPE_CACHE_IS_UPDATING(sequence) ((sequence) & 0x01)
```

colesbury commented on Objects/typeobject.c:
```diff
@@ -4747,6 +4877,82 @@ is_dunder_name(PyObject *name)
+static void
+update_cache(struct type_cache_entry *entry, PyObject *name, unsigned int version_tag, PyObject *value)
+{
+    entry->version = version_tag;
+    entry->value = value;  /* borrowed */
```
We will want to use relaxed atomic writes for version, value, and name because reads can happen concurrently.

colesbury commented on Objects/typeobject.c:
```diff
@@ -780,8 +821,8 @@ PyType_Unwatch(int watcher_id, PyObject* obj)
-void
-PyType_Modified(PyTypeObject *type)
+static void
+type_modified_unlocked(PyTypeObject *type)
```
This should call `type_modified_unlocked(subclass)` instead of `PyType_Modified(subclass)`.

Should this function also assert that the type lock is held?

(GitHub annoyingly won't let me comment on the actual line because it's too far away from a modified line)

colesbury commented on Objects/typeobject.c:
```diff
@@ -1177,21 +1241,28 @@ type_set_abstractmethods(PyTypeObject *type, PyObject *value, void *context)
-    PyObject *mro = lookup_tp_mro(type);
+    PyObject *mro;
+
+    BEGIN_TYPE_LOCK()
+    mro = lookup_tp_mro(type);
     if (mro == NULL) {
-        Py_RETURN_NONE;
+        mro = Py_None;
+    } else {
+        Py_INCREF(mro);
     }
-    return Py_NewRef(mro);
+
+    END_TYPE_LOCK()
+    return mro;
```
Maybe delegate to `_PyType_GetMro(type)`? Like `type_get_bases` does with `_PyType_GetBases` above.

colesbury commented on Objects/typeobject.c:
```diff
@@ -4747,6 +4877,82 @@ is_dunder_name(PyObject *name)
+    entry->value = value;  /* borrowed */
+    assert(_PyASCIIObject_CAST(name)->hash != -1);
+    OBJECT_STAT_INC_COND(type_cache_collisions, entry->name != Py_None && entry->name != name);
+    Py_SETREF(entry->name, Py_NewRef(name));
```
The implicit `Py_DECREF` under a lock (the seqlock) worries me, but I guess it's fine here because `name` must be a `PyUnicodeObject` (and not a subclass).

DinoV replied:
We could always return it and decref it after the lock, but I'll just add a comment to make it clear that it's depending on being an exact str right now.

colesbury commented on Objects/typeobject.c:
```diff
@@ -4759,20 +4965,59 @@ _PyType_Lookup(PyTypeObject *type, PyObject *name)
+        has_version = assign_version_tag(interp, type);
+        version = type->tp_version_tag;
+    }
+    END_TYPE_LOCK()
```
Should we be updating the cache below while still holding type lock?

DinoV replied:
It's fine that we're not doing that because we're capturing the version here with the lock held.  If something mutates the lock after this then we'll just stuff an invalid version entry into the cache versus having the cache entry get invalidated right after we release the lock.

colesbury commented on Objects/typeobject.c:
```diff
@@ -4789,16 +5034,14 @@ _PyType_Lookup(PyTypeObject *type, PyObject *name)
+#else
+        update_cache(entry, name, version, res);
+#endif
         assert(_PyType_HasFeature(type, Py_TPFLAGS_VALID_VERSION_TAG));
```
This assert is not safe here, at least in the free-threaded build because it's outside the type lock. Some other thread may have invalidating the version tag.

colesbury commented on Objects/typeobject.c:
```diff
@@ -9233,7 +9515,7 @@ slot_bf_getbuffer(PyObject *self, Py_buffer *buffer, int flags)
 
 static int
-releasebuffer_maybe_call_super(PyObject *self, Py_buffer *buffer)
+releasebuffer_maybe_call_super_unlocked(PyObject *self, Py_buffer *buffer, releasebufferproc *base_releasebuffer)
```
I think this would be easier to understand if it returned a `releasebufferproc` (possibly NULL) instead of coordinating the error return code with the pointer write.

The `-1` error return doesn't do anything because we never set a Python error.

DinoV replied:
Yeah weird, it doesn't look like it was discussed in the original PR.  I'll just get rid of the unused error handling stuff.

colesbury commented on Objects/typeobject.c:
```diff
@@ -9264,15 +9546,28 @@ releasebuffer_maybe_call_super(PyObject *self, Py_buffer *buffer)
+    res = releasebuffer_maybe_call_super_unlocked(self, buffer, &base_releasebuffer);
+    END_TYPE_LOCK();
+
+    if (res == 0) {
```
Related to the above comment about the `releasebuffer_maybe_call_super_unlocked` signature: This doesn't handle the case where `res == 0` but `base_releasebuffer` is `NULL`.

colesbury reviewed (commented):
The overall approach looks good.

I think the seqlock code would be easier to understand if it's refactored out into its own type with associated functions (possibly in `pycore_lock.h`).

DinoV committed:
Move seqlock implementation to lock.c

DinoV committed:
Review feedback

colesbury commented on Include/internal/pycore_lock.h:
```diff
@@ -251,6 +251,39 @@ PyAPI_FUNC(void) _PyRWMutex_RUnlock(_PyRWMutex *rwmutex);
+// The writer can also detect that the undelering data has not changed and abandon the write
+// and restore the previous sequence.
+typedef struct {
+    int sequence;
```
I think we should use an unsigned type (either `unsigned int` or `uint32_t`) for the sequence because it may wrap around.

DinoV replied:
The main reason I did signed was we're missing `_Py_atomic_load_uint32_acquire`, but I'll stop being lazy and add it :)

colesbury commented on Objects/typeobject.c:
```diff
@@ -51,6 +52,34 @@ class object "PyObject *" "&PyBaseObject_Type"
+// on the subclass lock.
+#define BEGIN_TYPE_LOCK()                                           \
+    {                                                               \
+    _PyCriticalSection _cs;                                         \
```
The indentation here doesn't match `END_TYPE_LOCK()`.

colesbury commented on Python/lock.c:
```diff
@@ -459,3 +459,72 @@ _PyRWMutex_Unlock(_PyRWMutex *rwmutex)
+{
+    int new_seq = seqlock->sequence - 1;
+    assert(!SEQLOCK_IS_UPDATING(new_seq));
+    _Py_atomic_exchange_int32(&seqlock->sequence, new_seq);
```
This can be an atomic store instead of an exchange.

colesbury commented on Python/lock.c:
```diff
@@ -459,3 +459,72 @@ _PyRWMutex_Unlock(_PyRWMutex *rwmutex)
+void _PySeqLock_LockWrite(_PySeqLock *seqlock)
+{
+    // lock the entry by setting by moving to an odd sequence number
+    int prev = seqlock->sequence;
```
This should be a relaxed read like you do in the loop below.

colesbury commented on Python/lock.c:
```diff
@@ -459,3 +459,72 @@ _PyRWMutex_Unlock(_PyRWMutex *rwmutex)
+            // Someone else is currently updating the cache
+            _Py_yield();
+            prev = _Py_atomic_load_int32_relaxed(&seqlock->sequence);
+        } else if (_Py_atomic_compare_exchange_int32(&seqlock->sequence, &prev, prev + 1)) {
```
style: `else if` on new line

colesbury commented on Python/lock.c:
```diff
@@ -459,3 +459,72 @@ _PyRWMutex_Unlock(_PyRWMutex *rwmutex)
+        } else if (_Py_atomic_compare_exchange_int32(&seqlock->sequence, &prev, prev + 1)) {
+            // We've locked the cache
+            break;
+        } else {
```
style: `else` on new line

colesbury commented on Python/lock.c:
```diff
@@ -459,3 +459,72 @@ _PyRWMutex_Unlock(_PyRWMutex *rwmutex)
+{
+    int new_seq = seqlock->sequence + 1;
+    assert(!SEQLOCK_IS_UPDATING(new_seq));
+    _Py_atomic_exchange_int32(&seqlock->sequence, new_seq);
```
This also can be an atomic store instead of an exchange.

colesbury reviewed (commented):
Looks good to me other than a few minor issues.

As I commented below, I think it'd be better to use an unsigned type for the sequence counter since it can wrap around. (We compile with `-fwrapv`, but still seems like good practice). Whatever type you use, the atomic operations should match. Currently, there's a mix of `_int32` and `_int` suffixed operations.

DinoV committed:
Use int32_t for sequence lock
Use atomic storage instead of exchange
Use relaxed load when starting sequence lock for write
Fix some formatting
